### PR TITLE
[FLINK-9476]Emit late elements in CEP as sideOutPut

### PR DESCRIPTION
**Pull request by Aitozi**
Now, when use with Eventtime in CEP library, elements come later than watermark will be dropped,we can put it in side Output with outPutTag

### CODE REVIEW

File: docs/dev/libs/cep.md [hunk 1]
```diff
@@ -1524,7 +1524,52 @@ In `CEP` the order in which elements are processed matters. To guarantee that el
...
-seen watermark. Late elements are not further processed.
+seen watermark. Late elements are not further processed. Also, you can specify a sideOutput tag to collect the late elements come after the last seen watermark, you can use it like this.
+
+<div class="codetabs" markdown="1">
+<div data-lang="java" markdown="1">
+
+{% highlight java %}
+PatternStream<Event> patternStream = CEP.pattern(input, pattern);
+
+OutputTag<String> lataDataOutputTag = new OutputTag<String>("lata-data""){};
```
Comment (bowenli86) [inline docs/dev/libs/cep.md]: typo: "lateDataOutputTag" and "late-data"

File: docs/dev/libs/cep.md [hunk 2]
```diff
@@ -1524,7 +1524,52 @@ In `CEP` the order in which elements are processed matters. To guarantee that el
...
+
+SingleOutputStreamOperator<ComplexEvent> result = patternStream
+    .sideOutputLateData(lataDataOutputTag)
+    .select(
+        new PatternTimeoutFunction<Event, TimeoutEvent>() {...},
+        outputTag,
+        new PatternSelectFunction<Event, ComplexEvent>() {...}
+    );
+
+DataStream<String> lataData = result.getSideOutput(lataDataOutputTag);
```
Comment (bowenli86) [inline docs/dev/libs/cep.md]: typo: lateData

**Comment by Aitozi** [comment]
@bowenli86 thanks for review, i have fix the error according to the comment and add the unit test in CEPITCase, please help review it again. cc @kl0u

File: flink-libraries/flink-cep/src/test/java/org/apache/flink/cep/CEPITCase.java [hunk 3]
```diff
@@ -383,6 +386,100 @@ public String select(Map<String, List<Event>> pattern) {
 		env.execute();
 	}
 
+	@Test
+	public void testSimpleKeyedPatternEventTimeWithSideOutput() throws Exception {
```
Comment (dawidwys) [inline flink-libraries/flink-cep/src/test/java/org/apache/flink/cep/CEPITCase.java]: How about to simplify the test a bit? 

I would be in favour of implementing this test with `OperatorTestHarness` (in CepOperatorTest). This way we it is way easier to follow which events are late.

Also I would use a simpler Pattern here... Don't see a reason for using 3 part one.

File: docs/dev/libs/cep.md [hunk 4]
```diff
@@ -1524,7 +1524,52 @@ In `CEP` the order in which elements are processed matters. To guarantee that el
...
+<div data-lang="scala" markdown="1">
+
+{% highlight scala %}
+val patternStream: PatternStream[Event] = CEP.pattern(input, pattern)
+
+val lateDataOutputTag = OutputTag[String]("late-data")
+
+val result: SingleOutputStreamOperator[ComplexEvent] = patternStream
+    .sideOutputLateData(lateDataOutputTag)
+    .select(outputTag){
```
[the remark this answers is not in the document]
Comment (Aitozi) [reply docs/dev/libs/cep.md]: OK, I remove the timeout tag to reduce confusion, and fix a little error in before doc(the sequence of the parameter is wrong)

File: docs/dev/libs/cep.md [hunk 5]
```diff
@@ -1524,7 +1524,52 @@ In `CEP` the order in which elements are processed matters. To guarantee that el
...
+
+<div class="codetabs" markdown="1">
+<div data-lang="java" markdown="1">
+
+{% highlight java %}
+PatternStream<Event> patternStream = CEP.pattern(input, pattern);
+
+OutputTag<String> lateDataOutputTag = new OutputTag<String>("late-data""){};
+
+OutputTag<String> outputTag = new OutputTag<String>("side-output""){};
```
[the remark this answers is not in the document]
Comment (Aitozi) [reply docs/dev/libs/cep.md]: removed

File: flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/PatternStream.java [hunk 6]
```diff
@@ -54,6 +54,12 @@
 	// comparator to sort events
 	private final EventComparator<T> comparator;
 
+	/**
+	 * Side output {@code OutputTag} for late data. If no tag is set late data will simply be
```
[the remark this answers is not in the document]
Comment (Aitozi) [reply flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/PatternStream.java]: fixed

File: flink-libraries/flink-cep/src/test/java/org/apache/flink/cep/CEPITCase.java (hunk 3 above)
Comment (Aitozi) [reply flink-libraries/flink-cep/src/test/java/org/apache/flink/cep/CEPITCase.java]: move the test case to CepOperatorTest, please help review again, thanks ;)

**Comment by dawidwys** [comment]
+1, LGTM

**Comment by Aitozi** [comment]
thx, can help merge this @dawidwys

**Comment by dawidwys** [comment]
I will merge it either later today or tomorrow. Thanks once again!

File: docs/dev/libs/cep.md [hunk 7]
```diff
@@ -1524,7 +1524,48 @@ In `CEP` the order in which elements are processed matters. To guarantee that el
...
+      .sideOutputLateData(lateDataOutputTag)
+      .select(
+        {pattern: Map[String, Iterable[ComplexEvent]] => ComplexEvent()
+      })
+
+val lateData: DataStream<String> = result.getSideOutput(lateDataOutputTag)
+
+{% endhighlight %}
+
+</div>
```
Comment (dawidwys) [inline docs/dev/libs/cep.md]: You missed closing `</div>`

File: docs/dev/libs/cep.md [hunk 8]
```diff
@@ -1524,7 +1524,48 @@ In `CEP` the order in which elements are processed matters. To guarantee that el
...
+
+{% highlight scala %}
+
+val patternStream: PatternStream[Event] = CEP.pattern(input, pattern)
+
+val lateDataOutputTag = OutputTag[String]("late-data")
+
+val result: SingleOutputStreamOperator[ComplexEvent] = patternStream
+      .sideOutputLateData(lateDataOutputTag)
+      .select(
```
Comment (dawidwys) [inline docs/dev/libs/cep.md]: IMO this is not a natural scala construct.
Could you switch to either:

    .select{
        pattern: Map[String, Iterable[ComplexEvent]] => ComplexEvent()
    }

or

    .select((pattern: Map[String, Iterable[ComplexEvent]]) => ComplexEvent())

File: flink-libraries/flink-cep/src/test/java/org/apache/flink/cep/operator/CEPOperatorTest.java [hunk 9]
```diff
@@ -726,6 +727,50 @@ public void testCEPOperatorCleanupEventTimeWithSameElements() throws Exception {
...
+		Event middle1Event1 = new Event(41, "a", 2.0);
+		Event middle1Event2 = new Event(41, "a", 3.0);
+		Event middle1Event3 = new Event(41, "a", 4.0);
+
+		SelectCepOperator<Event, Integer, Map<String, List<Event>>> operator = CepOperatorTestUtilities.getKeyedCepOpearator(
+			false,
+			new ComplexNFAFactory());
+		OneInputStreamOperatorTestHarness<Event, Map<String, List<Event>>> harness = CepOperatorTestUtilities.getCepTestHarness(operator);
+
+		try {
```
Comment (dawidwys) [inline flink-libraries/flink-cep/src/test/java/org/apache/flink/cep/operator/CEPOperatorTest.java]: Could you switch it to `try-with-resources` block? I know there is a lot tests in that style, but we should add new tests with a proper style ;)

File: flink-libraries/flink-cep/src/test/java/org/apache/flink/cep/operator/CepOperatorTestUtilities.java [hunk 10]
```diff
@@ -66,6 +68,9 @@ public Integer getKey(Event value) throws Exception {
 		boolean isProcessingTime,
 		NFACompiler.NFAFactory<Event> nfaFactory,
 		EventComparator<Event> comparator) {
+
+		OutputTag<Event> lateDataTag = new OutputTag<Event>("late-data", TypeInformation.of(Event.class));
```
Comment (dawidwys) [inline flink-libraries/flink-cep/src/test/java/org/apache/flink/cep/operator/CepOperatorTestUtilities.java]: Could you pass it as parameter? And by default set a null one? I would prefer that we do not depend on the fact that the name in here and in the test are the same. Would prefer using the same object.

**Comment by Aitozi** [comment]
@dawidwys fix the flaw according to your suggestion ;-)

**Pull request closed without merging**